canine/orchestrator.py

```diff
@@ -463,119 +527,126 @@ def make_output_DF(self, batch_id, outputs, cpu_time, prev_acct, localizer = Non
                     pandas_write_hdf5_buffered(df, buf = w, key = "results")
         return df
 
-    def submit_batch_job(self, entrypoint_path, compute_env, extra_sbatch_args = {}) -> int:
+    def submit_batch_job(self, entrypoint_path, compute_env, extra_sbatch_args = {}, job_spec = None) -> int:
+        if job_spec is None:
+            job_spec = self.job_spec
+
         # this job was avoided
-        if len(self.job_spec) == 0:
+        if len(job_spec) == 0:
             return -2
 
         batch_id = self.backend.sbatch(
             entrypoint_path,
+            "H", # submit in "held" state to allow for noop'd jobs to be cancelled
             **{
                 'requeue': True,
                 'job_name': self.name,
-                'array': "0-{}".format(len(self.job_spec)-1),
+                'array': "0-{}".format(len(job_spec)-1),
                 'output': "{}/%a/stdout".format(compute_env['CANINE_JOBS']),
                 'error': "{}/%a/stderr".format(compute_env['CANINE_JOBS']),
                 **self.resources,
                 **Orchestrator.stringify(extra_sbatch_args)
             }
         )
 
+        # cancel noop'd jobs
+        for k, v in job_spec.items():
+            if v is None:
+                self.backend.scancel(batch_id + "_" + k)
```

Comment:
Will avoided jobs show up as cancelled in the `sacct` table?

Reply:
Yup! That's the whole point of using noop'd jobs instead of just not dispatching them in the first place.

Reply:
I think this is great for now, but in the future, we might be able to take this a step further and exercise granular control of the job shards that actually get dispatched by using `--array`'s ability to comprehend complex ranges. Like we could just avoid scheduling avoided jobs in the first place. Not sure how you feel about that idea

Reply:
I'll add this as a feature issue. Turns out that job avoidance on massive jobs is slow, since each `scancel` takes a nontrivial amount of time. Fix #1 would be using range expressions for `scancel`; if that speeds things up, I'd leave it at that. Fix #2 (your proposed solution using `sbatch --array ...`) would require a little extra hacking on Canine's end, since we'd then have to manually add rows for avoided jobs to the accounting dataframe.